src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -3662,6 +3662,149 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
   bind(L_end);
 }
 
+static const int64_t bits32 = right_n_bits(32);
+static const int64_t bits8 = right_n_bits(8);
+
+/**
+ * Emits code to update CRC-32 with a byte value according to constants in table
+ *
+ * @param [in,out]crc   Register containing the crc.
+ * @param [in]val       Register containing the byte to fold into the CRC.
+ * @param [in]table     Register containing the table of crc constants.
+ *
+ * uint32_t crc;
+ * val = crc_table[(val ^ crc) & 0xFF];
+ * crc = val ^ (crc >> 8);
+ *
+ */
+void MacroAssembler::update_byte_crc32(Register crc, Register val, Register table) {
+  assert_different_registers(crc, val, table);
+
+  xorr(val, val, crc);
+  andi(val, val, bits8);
+  slli(val, val, 2);
+  add(val, table, val);
+  Assembler::lwu(val, val, 0);
+  srli(crc, crc, 8);
+  xorr(crc, val, crc);
+}
+
+/**
+ * Emits code to update CRC-32 with a 32-bit value according to tables 0 to 3
+ *
+ * @param [in,out]crc   Register containing the crc.
+ * @param [in]v         Register containing the 32-bit to fold into the CRC.
+ * @param [in]table0    Register containing table 0 of crc constants.
+ * @param [in]table1    Register containing table 1 of crc constants.
+ * @param [in]table2    Register containing table 2 of crc constants.
+ * @param [in]table3    Register containing table 3 of crc constants.
+ *
+ * uint32_t crc;
+ *   v = crc ^ v
+ *   crc = table3[v&0xff]^table2[(v>>8)&0xff]^table1[(v>>16)&0xff]^table0[v>>24]
+ *
+ */
+void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1, Register tmp2,
+        Register table0, Register table1, Register table2, Register table3, bool upper) {
+  assert_different_registers(crc, v, tmp1, tmp2, table0, table1, table2, table3);
+
+  if (upper)
+    srli(v, v, 32);
+  xorr(v, crc, v);
+
+  andi(tmp1, v, bits8);
+  shadd(tmp1, tmp1, table3, tmp2, 2);
+  Assembler::lwu(crc, tmp1, 0);
```

Comment:
Why not use `MacroAssembler::lwu` instead ? I see low difference in stub code emitted.
Like:
``` diff
diff --git a/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp b/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp
index 06026b98bfa..eb9362ca531 100644
--- a/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp
+++ b/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp
@@ -3696,26 +3696,26 @@ void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1,
 
   andi(tmp1, v, bits8);
   shadd(tmp1, tmp1, table3, tmp2, 2);
-  Assembler::lwu(crc, tmp1, 0);
+  lwu(crc, Address(tmp1, 0));
 
   srli(tmp1, v, 6);
   andi(tmp1, tmp1, (bits8 << 2));
   add(tmp1, tmp1, table2);
-  Assembler::lwu(tmp2, tmp1, 0);
+  lwu(tmp2, Address(tmp1, 0));
 
   srli(tmp1, v, 14);
   xorr(crc, crc, tmp2);
 
   andi(tmp1, tmp1, (bits8 << 2));
   add(tmp1, tmp1, table1);
-  Assembler::lwu(tmp2, tmp1, 0);
+  lwu(tmp2, Address(tmp1, 0));
 
   srli(tmp1, v, 22);
   xorr(crc, crc, tmp2);
 
   andi(tmp1, tmp1, (bits8 << 2));
   add(tmp1, tmp1, table0);
-  Assembler::lwu(tmp2, tmp1, 0);
+  lwu(tmp2, Address(tmp1, 0));
   xorr(crc, crc, tmp2);
 }
```

Reply:
See https://github.com/openjdk/jdk/pull/17122#issuecomment-1862404200

Reply:
When I tried `MacroAssembler::lwu` I got the following instructions on T-head:
```
0.47%  ?  0x0000003fac6a8738:   li	t3,1
0.51%  ?  0x0000003fac6a873a:   slli	t3,t3,0x20
0.00%  ?  0x0000003fac6a873c:   addi	t3,t3,-1
...
2.68%  ?  0x0000003fac6a8752:   lw	a0,0(t1)
5.25%  ?  0x0000003fac6a8756:   and	a0,a0,t3
...
       ?  0x0000003fac6a876a:   lw	t4,0(t1)
1.78%  ?  0x0000003fac6a876e:   and	t1,t4,t3
...
0.49%  ?  0x0000003fac6a8786:   lw	t4,0(t1)
2.62%  ?  0x0000003fac6a878a:   and	t1,t4,t3
...
0.41%  ?  0x0000003fac6a87a2:   lw	t4,0(t1)
3.97%  ?  0x0000003fac6a87a6:   and	t1,t4,t3
```
instead of just 
```
4.52%  ??  0x0000003fb49e96f6:   lwu	a0,0(t1)
...
       ??  0x0000003fb49e970a:   lwu	t3,0(t1)
...
       ??  0x0000003fb49e9722:   lwu	t3,0(t1)
...
0.02%  ??  0x0000003fb49e973a:   lwu	t3,0(t1)
```

Reply:
Interesting, I tried on qemu and `T-HEAD Light Lichee Pi 4A`, I don't get this code generated with just `lwu(tmp2, Address(tmp1, 0));`.
Do you know how does it happen?

Reply:
Whoops, I tried to reproduce behavior above and it turned out that it was just some old profiling data I collected before implementing version with `lwu` (previously I used the `lw` + `andi` pair of instructions). The `MacroAssembler::lwu` version actually works fine. Thanks everyone for pointing it out!